### PR TITLE
Add support for no-op concat with padded output

### PR DESCRIPTION
**Pull request by nandesuka**
Add support for no-op concat with padded output

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

### CODE REVIEW

File: test/inductor/test_torchinductor_codegen_config_overrides.py [hunk 1]
```diff
@@ -108,6 +108,50 @@ def func(a, b):
...
         )
         self.count_code("@triton.jit", code, 3)
+    
+    @parametrize("shape,alignment_bytes", 
+        [
+        ((512, 1), 32), 
+        ((32, 30), 64), 
+        ]
+    )
+    def test_concat_alignments(self, shape, alignment_bytes):
```
Comment (blaine-rister) [inline test/inductor/test_torchinductor_codegen_config_overrides.py:118]: Nice test!

File: test/inductor/test_torchinductor_codegen_config_overrides.py [hunk 2]
```diff
@@ -108,6 +108,50 @@ def func(a, b):
...
+                "pad_channels_last": True,
+                "pad_outputs": True,
+                "padding_stride_threshold": 0,
+                "triton.prefer_nd_tiling": True,
+                "max_pointwise_cat_inputs" : 1,
+        })
+       
+        output_shape = (shape[0] + shape[0], shape[1])
+        output_stride = x.stride()
+        output_line = f"buf3 = empty_strided_cpu({output_shape}, {output_stride}, torch.float32)"
```
Comment (blaine-rister) [inline test/inductor/test_torchinductor_codegen_config_overrides.py:153]: If the output code is this specific, you may want to consider using `expecttest` asserts. Those are a bit easier to maintain because you can automatically update the code with `EXPECTTEST_ACCEPT=1`. Here's an example. https://github.com/pytorch/pytorch/blob/main/test/inductor/test_torchinductor_strided_blocks.py#L350

More info on `expecttest`:
https://github.com/pytorch/expecttest

File: torch/_inductor/ir.py [hunk 3]
```diff
@@ -4634,6 +4634,8 @@ def create(cls, inputs, dim):  # type: ignore[no-untyped-def]
             offsets_end.append(new_size[dim])
 
         output_stride = FlexibleLayout.contiguous_strides(new_size)
+        # Ensure the output stride matches the alignment requirements
+        output_stride = Layout._pad_strides(output_stride, new_size, inputs[0].dtype)
```
Comment (blaine-rister) [inline torch/_inductor/ir.py:4638]: I'm wondering if padded concats have any perf drawbacks. It seems like padding could be beneficial on GPUs to avoid bank conflicts, but I'm not sure if there's a reason the existing code specifically asks for contiguous strides. It might be good to double check for perf regressions by running some benchmarks.

You can kick off an A100 perf run here. https://github.com/pytorch/pytorch/actions/workflows/inductor-perf-test-nightly.yml

It might also be useful to write a microbenchmark showing whether this impacts perf on programs that only contain concats. If you run Inductor in debug mode, it will automatically generate a benchmarking script. https://pytorch.org/tutorials/intermediate/inductor_debug_cpu.html#get-more-logging-information

cc @shunting314  @eellison

**Review by blaine-rister — COMMENTED** [review]
This mostly LGTM. I left a few small nits.

Before accepting, it would be good to see some GPU benchmarks to confirm this is at least perf neutral. (It might actually be perf positive if it avoids bank conflicts.)

cc @shunting314  in case you want to suggest some representative test programs.

File: test/inductor/test_torchinductor_codegen_config_overrides.py (hunk 1 above)
Comment (blaine-rister) [reply test/inductor/test_torchinductor_codegen_config_overrides.py:118]: Can I suggest moving this test to this module? https://github.com/pytorch/pytorch/blob/main/test/inductor/test_padding.py

That contains a lot of similar padding tests.

File: torch/_inductor/ir.py [hunk 4]
```diff
@@ -4634,6 +4634,8 @@ def create(cls, inputs, dim):  # type: ignore[no-untyped-def]
             offsets_end.append(new_size[dim])
 
         output_stride = FlexibleLayout.contiguous_strides(new_size)
```
[the remark this answers is not in the document]
Comment (blaine-rister) [reply torch/_inductor/ir.py:4636]: On second thought, it seems like we should choose whether to pad based on `config.comprehensive_padding`. That would be more consistent with how `FlexibleLayout` usually chooses padding.

File: test/inductor/test_torchinductor_codegen_config_overrides.py [hunk 5]
```diff
@@ -108,6 +108,50 @@ def func(a, b):
...
+                "padding_alignment_bytes": alignment_bytes,
+                "disable_padding_cpu": False,
+                "pad_channels_last": True,
+                "pad_outputs": True,
+                "padding_stride_threshold": 0,
+                "triton.prefer_nd_tiling": True,
+                "max_pointwise_cat_inputs" : 1,
+        })
+       
+        output_shape = (shape[0] + shape[0], shape[1])
```
Comment (blaine-rister) [inline test/inductor/test_torchinductor_codegen_config_overrides.py:151]: Can we also test what happens when `config.pad_outputs == False`? AFAIK Inductor will try to match eager mode strides (i.e., not pad) when that happens.

File: torch/_inductor/ir.py (hunk 3 above)
Comment (eellison) [reply torch/_inductor/ir.py:4638]: If we have a no-op concat like:
```
def foo(a, b, x, y):
     return torch.cat((a @ b, x @ y))
```

Then padding the stride here I think will be perf-negative, bc we will no longer be able to plan this. The inputs were dense but our stride is not.

I think we want to do some checking to avoid padding in this case.

File: torch/_inductor/ir.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (eellison) [reply torch/_inductor/ir.py:4636]: Also, as @blaine-rister says youre skipping the `should_pad_strides` path

**Review by eellison — COMMENTED** [review]
cc @shunting314 for thoughts too

**Comment by shunting314** [comment]
Looks good to me. +1 to guard the code behind the config.comprehensive_padding .

File: torch/_inductor/ir.py (hunk 3 above)
Comment (nandesuka) [reply torch/_inductor/ir.py]: Perf run completed. https://github.com/pytorch/pytorch/actions/runs/13567211431

**Review by shunting314 — APPROVED** [review]
